File: packages/gdanav_app/android/src/main/kotlin/it/gdanav/gdanav_app/auto/PonteAuto.kt
package it.gdanav.gdanav_app.auto

import android.content.Context
import android.content.SharedPreferences
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.os.Handler
import android.os.Looper
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.util.concurrent.CopyOnWriteArrayList

/**
 * Quello che il telefono sa e lo schermo dell'auto mostra. L'app Flutter lo
 * aggiorna sul canale `gdanav/schermo_auto`; l'auto manda indietro batteria e
 * autonomia sul canale `gdanav/auto`, e «fine» quando si ferma la guida.
 */
object PonteAuto {
    /** La prossima manovra e il viaggio, mentre si guida. */
    data class Guida(
        val tipo: Int,
        val distanzaM: Double,
        val strada: String,
        val istruzione: String,
        val restantiM: Double,
        val restantiS: Long,
        val arrivoMs: Long,
        val destinazione: String,
        val uscita: String = "",
        val verso: String = "",
        val rotonda: Int? = null,
        val corsie: List<CorsiaAuto> = emptyList(),
        val dopoTipo: Int? = null,
        val dopoStrada: String = "",
        /** La vista dello svincolo da mostrare, se pronta (vedi [svincoli]). */
        val svincolo: Int? = null,
    )

    /** Una corsia prima dello svincolo: le frecce, se è giusta, quale seguire. */
    data class CorsiaAuto(val direzioni: List<String>, val giusta: Boolean, val consigliata: String?)

    @Volatile var stileChiaro: String? = null
    @Volatile var stileScuro: String? = null

    /** I GeoJSON delle sorgenti dello stile, come li usa la mappa del telefono. */
    @Volatile var sorgenti: Map<String, String> = emptyMap()

    /** Dove sei: centro della mappa, e la rotta in gradi. */
    @Volatile var qui: DoubleArray? = null
    @Volatile var rotta: Double = 0.0
    @Volatile var guida: Guida? = null

    /** Un posto da scegliere in auto: Casa, Lavoro, un preferito, un recente o un risultato. */
    data class Luogo(
        val etichetta: String,
        val nome: String,
        val descrizione: String,
        val lat: Double,
        val lon: Double,
        val tipo: String,
    ) {
        fun comeMappa(): Map<String, Any> =
            mapOf("nome" to nome, "descrizione" to descrizione, "lat" to lat, "lon" to lon)
    }

    @Volatile var luoghi: List<Luogo> = emptyList()

    fun casa(): Luogo? = luoghi.firstOrNull { it.tipo == "casa" }
    fun lavoro(): Luogo? = luoghi.firstOrNull { it.tipo == "lavoro" }

    /** Sopra la mappa: batteria, velocità e limite, arrivo, sosta, meteo. */
    data class Cruscotto(
        val batteria: Double? = null,
        val autonomiaKm: Double? = null,
        /** L'autonomia la dice l'auto (non una stima). */
        val autonomiaAuto: Boolean = false,
        val velocita: Double? = null,
        val limite: Int? = null,
        val arrivoBatteria: Double? = null,
        val sostaNome: String? = null,
        val sostaKm: Double? = null,
        val sostaBatteria: Double? = null,
        val meteoTemperatura: Double? = null,
        val meteoEmoji: String? = null,
        val meteoDove: String? = null,
    )

    @Volatile var cruscotto = Cruscotto()

    /** La segnalazione che si avvicina, e quella appena passata. */
    data class Avviso(
        val titolo: String? = null,
        val tipo: String? = null,
        val metri: Double? = null,
        val limite: Int? = null,
        val ancoraId: String? = null,
        val ancoraTesto: String? = null,
    )

    @Volatile var avviso = Avviso()

    /** Le opzioni del percorso e la voce, per il menu. */
    @Volatile var opzioni: Map<String, Any?> = emptyMap()

    /** Le viste degli svincoli, disegnate dall'app: le ultime, per manovra. */
    @Volatile var svincoli: Map<Int, Bitmap> = emptyMap()

    /** Le icone delle segnalazioni, disegnate dall'app. */
    @Volatile var immagini: Map<String, Bitmap> = emptyMap()

    /** Cosa dire quando non si guida: «Calcolo il percorso…», o cosa non va. */
    @Volatile var messaggio: String? = null

    /** gdanav Premium: l'app lo dice, l'auto lo ricorda anche a telefono spento. */
    private var preferenze: SharedPreferences? = null

    fun premium(context: Context): Boolean =
        context.getSharedPreferences("gdanav", Context.MODE_PRIVATE).getBoolean("premium", false)

    private val principale = Handler(Looper.getMainLooper())
    private val ascoltatori = CopyOnWriteArrayList<() -> Unit>()
    private var canale: MethodChannel? = null
    private var sinkEnergia: EventChannel.EventSink? = null
    private var ultimaEnergia: Map<String, Any?>? = null

    fun ascolta(f: () -> Unit) = ascoltatori.add(f)

    fun smetti(f: () -> Unit) = ascoltatori.remove(f)

    fun collega(messenger: BinaryMessenger, context: Context) {
        preferenze = context.applicationContext.getSharedPreferences("gdanav", Context.MODE_PRIVATE)
        canale = MethodChannel(messenger, "gdanav/schermo_auto").also { c ->
            c.setMethodCallHandler { call, risultato ->
                gestisci(call)
                risultato.success(null)
            }
        }
        EventChannel(messenger, "gdanav/auto").setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink) {
                sinkEnergia = events
                if (ultimaEnergia != null || ultimaVelocita != null) mandaAuto()
            }

            override fun onCancel(arguments: Any?) {
                sinkEnergia = null
            }
        })
    }

    /**
     * Cresce quando cambia qualcosa dei modelli di Android Auto (manovra,
     * messaggi, luoghi, opzioni): il resto (posizione, cruscotto) ridisegna
     * solo la mappa, senza consumare gli aggiornamenti concessi dall'auto.
     */
    @Volatile var versioneModello = 0
        private set

    private val soloMappa = setOf("posizione", "sorgenti", "cruscotto", "immagini", "stili")

    private fun gestisci(call: MethodCall) {
        val prima = avviso.ancoraId
        when (call.method) {
            "stili" -> {
                stileChiaro = call.argument("chiaro")
                stileScuro = call.argument("scuro")
            }
            "sorgenti" -> {
                val nuove = HashMap(sorgenti)
                call.argument<Map<String, String>>("dati")?.let { nuove.putAll(it) }
                sorgenti = nuove
            }
            "posizione" -> {
                val lat = call.argument<Double>("lat")
                val lon = call.argument<Double>("lon")
                qui = if (lat != null && lon != null) doubleArrayOf(lat, lon) else null
                rotta = call.argument<Double>("rotta") ?: 0.0
            }
            "luoghi" -> {
                luoghi = (call.argument<List<Map<String, Any?>>>("elenco") ?: emptyList()).mapNotNull(::luogo)
            }
            "messaggio" -> messaggio = call.argument<String>("testo")
            "cruscotto" -> cruscotto = Cruscotto(
                batteria = numero(call, "batteria"),
                autonomiaKm = numero(call, "autonomia_km"),
                autonomiaAuto = call.argument<Boolean>("autonomia_auto") == true,
                velocita = numero(call, "velocita"),
                limite = numero(call, "limite")?.toInt(),
                arrivoBatteria = numero(call, "arrivo_batteria"),
                sostaNome = call.argument<String>("sosta_nome"),
                sostaKm = numero(call, "sosta_km"),
                sostaBatteria = numero(call, "sosta_batteria"),
                meteoTemperatura = numero(call, "meteo_temperatura"),
                meteoEmoji = call.argument<String>("meteo_emoji"),
                meteoDove = call.argument<String>("meteo_dove"),
            )
            "avviso" -> avviso = Avviso(
                titolo = call.argument<String>("titolo"),
                tipo = call.argument<String>("tipo"),
                metri = numero(call, "metri"),
                limite = numero(call, "limite")?.toInt(),
                ancoraId = call.argument<String>("ancora_id"),
                ancoraTesto = call.argument<String>("ancora_testo"),
            )
            "opzioni" -> opzioni = (call.arguments as? Map<*, *>)?.entries?.associate { "${it.key}" to it.value } ?: emptyMap()
            "svincolo" -> {
                val id = numero(call, "id")?.toInt()
                val png = call.argument<ByteArray>("png")
                val bitmap = png?.let { BitmapFactory.decodeByteArray(it, 0, it.size) }
                if (id != null && bitmap != null) {
                    svincoli = svincoli.entries.toList().takeLast(2).associate { it.key to it.value } + (id to bitmap)
                }
            }
            "immagini" -> {
                val nuove = HashMap(immagini)
                (call.argument<Map<String, ByteArray>>("png") ?: emptyMap()).forEach { (nome, byte) ->
                    BitmapFactory.decodeByteArray(byte, 0, byte.size)?.let { nuove[nome] = it }
                }
                immagini = nuove
            }
            "premium" -> preferenze?.edit()?.putBoolean("premium", call.argument<Boolean>("sbloccato") == true)?.apply()
            "guida" -> {
                guida = if (call.argument<Boolean>("attiva") == true) {
                    Guida(
                        tipo = call.argument<Int>("tipo") ?: 8,
                        distanzaM = call.argument<Double>("distanza") ?: 0.0,
                        strada = call.argument<String>("strada") ?: "",
                        istruzione = call.argument<String>("istruzione") ?: "",
                        restantiM = call.argument<Double>("restanti") ?: 0.0,
                        restantiS = (call.argument<Number>("secondi") ?: 0).toLong(),
                        arrivoMs = (call.argument<Number>("arrivo") ?: System.currentTimeMillis()).toLong(),
                        destinazione = call.argument<String>("destinazione") ?: "",
                        uscita = call.argument<String>("uscita") ?: "",
                        verso = call.argument<String>("verso") ?: "",
                        rotonda = numero(call, "rotonda")?.toInt(),
                        corsie = (call.argument<List<Map<String, Any?>>>("corsie") ?: emptyList()).map { c ->
                            CorsiaAuto(
                                direzioni = (c["direzioni"] as? List<*>)?.map { "$it" } ?: emptyList(),
                                giusta = c["giusta"] == true,
                                consigliata = c["consigliata"] as? String,
                            )
                        },
                        dopoTipo = numero(call, "dopo_tipo")?.toInt(),
                        dopoStrada = call.argument<String>("dopo_strada") ?: "",
                        svincolo = numero(call, "svincolo")?.toInt(),
                    )
                } else {
                    null
                }
            }
        }
        if (call.method !in soloMappa && (call.method != "avviso" || avviso.ancoraId != prima)) versioneModello++
        avvisa()
    }

    private fun numero(call: MethodCall, chiave: String): Double? = (call.argument<Any?>(chiave) as? Number)?.toDouble()

    /** Una domanda all'app, con la risposta sul filo principale. */
    fun chiedi(metodo: String, argomenti: Any?, risposta: (Any?) -> Unit = {}) {
        val c = canale ?: return risposta(null)
        principale.post {
            c.invokeMethod(metodo, argomenti, object : MethodChannel.Result {
                override fun success(r: Any?) = risposta(r)

                override fun error(codice: String, messaggio: String?, dettagli: Any?) = risposta(null)

                override fun notImplemented() = risposta(null)
            })
        }
    }

    /** Casa o Lavoro scelti dalla ricerca sull'auto. */
    fun imposta(tipo: String, l: Luogo, fatto: (Boolean) -> Unit) =
        chiedi("imposta", mapOf("tipo" to tipo, "luogo" to l.comeMappa())) { fatto(it == true) }

    fun cambiaOpzione(chiave: String, valore: Any) = chiedi("opzioni", mapOf(chiave to valore))

    fun alternaVoce() = chiedi("voce", null)

    /** Segnala dove sei; la risposta è la frase da mostrare. */
    fun segnala(tipo: String, risposta: (String) -> Unit) =
        chiedi("segnala", mapOf("tipo" to tipo)) { risposta(it as? String ?: "Non è partita.") }

    fun ancora(id: String, si: Boolean) = chiedi("ancora", mapOf("id" to id, "si" to si))

    /** Le colonnine rapide vicine, come luoghi da raggiungere. */
    fun colonnine(risultati: (List<Luogo>) -> Unit) = chiedi("colonnine", null) { r ->
        @Suppress("UNCHECKED_CAST")
        risultati(((r as? List<Map<String, Any?>>) ?: emptyList()).mapNotNull { luogo(it + ("tipo" to "colonnina")) })
    }

    private fun luogo(m: Map<String, Any?>): Luogo? {
        val lat = (m["lat"] as? Number)?.toDouble() ?: return null
        val lon = (m["lon"] as? Number)?.toDouble() ?: return null
        val nome = m["nome"] as? String ?: return null
        return Luogo(
            etichetta = m["etichetta"] as? String ?: nome,
            nome = nome,
            descrizione = m["descrizione"] as? String ?: "",
            lat = lat,
            lon = lon,
            tipo = m["tipo"] as? String ?: "risultato",
        )
    }

    /** La ricerca fatta sull'auto, con Photon come sul telefono. */
    fun cerca(testo: String, risultati: (List<Luogo>) -> Unit) {
        val c = canale ?: return risultati(emptyList())
        c.invokeMethod("cerca", mapOf("testo" to testo), object : MethodChannel.Result {
            override fun success(r: Any?) {
                @Suppress("UNCHECKED_CAST")
                risultati(((r as? List<Map<String, Any?>>) ?: emptyList()).mapNotNull(::luogo))
            }

            override fun error(codice: String, messaggio: String?, dettagli: Any?) = risultati(emptyList())

            override fun notImplemented() = risultati(emptyList())
        })
    }

    /** Una meta scelta sull'auto: il telefono calcola e parte la guida. */
    fun vai(l: Luogo) {
        messaggio = "Calcolo il percorso per ${l.etichetta}…"
        versioneModello++
        avvisa()
        principale.post { canale?.invokeMethod("vai", l.comeMappa()) }
    }

    private fun avvisa() = principale.post { ascoltatori.forEach { it() } }

    private var ultimaVelocita: Double? = null
    private var ultimoOdometro: Double? = null

    /** Dall'auto: batteria in percentuale e autonomia in metri. */
    fun energiaDallAuto(batteria: Float, autonomiaM: Float?) {
        ultimaEnergia = mapOf(
            "batteria" to batteria.toDouble(),
            "autonomia_km" to autonomiaM?.let { it / 1000.0 },
        )
        mandaAuto()
    }

    /** Dall'auto: la velocità del cruscotto, in m/s. */
    fun velocitaDallAuto(ms: Float) {
        ultimaVelocita = ms * 3.6
        mandaAuto()
    }

    /** Dall'auto: il contachilometri, in metri. */
    fun contachilometriDallAuto(metri: Float) {
        ultimoOdometro = metri / 1000.0
        mandaAuto()
    }

    /** Tutto quello che l'auto ha detto finora, in una lettura sola. */
    private fun mandaAuto() {
        val dati = HashMap<String, Any?>(ultimaEnergia ?: emptyMap())
        dati["velocita_kmh"] = ultimaVelocita
        dati["odometro_km"] = ultimoOdometro
        dati["letto_ms"] = System.currentTimeMillis()
        principale.post { sinkEnergia?.success(dati) }
    }

    /** L'auto ha chiesto di finire la guida (tasto sullo schermo o assistente). */
    fun fermaDallAuto() = principale.post { canale?.invokeMethod("ferma", null) }
}

File: packages/gdanav_app/ios/gdanav_app/Sources/gdanav_app/PonteAuto.swift
import Flutter
import UIKit

/// Un posto da scegliere in auto: Casa, Lavoro, un preferito, un recente o un risultato.
struct LuogoAuto {
    let etichetta: String
    let nome: String
    let descrizione: String
    let lat: Double
    let lon: Double
    let tipo: String

    var comeMappa: [String: Any] {
        ["nome": nome, "descrizione": descrizione, "lat": lat, "lon": lon]
    }

    init?(_ m: [String: Any], tipo predefinito: String = "risultato") {
        guard let lat = numero(m["lat"]), let lon = numero(m["lon"]), let nome = m["nome"] as? String else {
            return nil
        }
        self.nome = nome
        self.lat = lat
        self.lon = lon
        etichetta = m["etichetta"] as? String ?? nome
        descrizione = m["descrizione"] as? String ?? ""
        tipo = m["tipo"] as? String ?? predefinito
    }
}

/// Una corsia prima dello svincolo: le frecce, se è giusta, quale seguire.
struct CorsiaAuto {
    let direzioni: [String]
    let giusta: Bool
    let consigliata: String?
}

/// La prossima manovra e il viaggio, mentre si guida.
struct GuidaAuto {
    let tipo: Int
    let distanzaM: Double
    let strada: String
    let istruzione: String
    let restantiM: Double
    let restantiS: Double
    let arrivo: Date
    let destinazione: String
    let destinazioneLat: Double?
    let destinazioneLon: Double?
    let arrivoBatteria: Double?
    let uscita: String
    let verso: String
    let rotonda: Int?
    let corsie: [CorsiaAuto]
    let dopoTipo: Int?
    let dopoStrada: String
    /// La vista dello svincolo da mostrare, se pronta (vedi `svincoli`).
    let svincolo: Int?
}

/// Batteria, velocità e limite, arrivo, sosta, meteo.
struct CruscottoAuto {
    /// Auto termica: niente batteria né colonnine.
    var elettrica = true
    var batteria: Double?
    var autonomiaKm: Double?
    var velocita: Double?
    var limite: Int?
    var arrivoBatteria: Double?
    var sostaNome: String?
    var sostaKm: Double?
    var sostaBatteria: Double?
    var meteoTemperatura: Double?
    var meteoEmoji: String?
    var meteoDove: String?
}

/// La segnalazione che si avvicina, e quella appena passata.
struct AvvisoAuto {
    var titolo: String?
    var tipo: String?
    var metri: Double?
    var limite: Int?
    var ancoraId: String?
    var ancoraTesto: String?
}

func numero(_ v: Any?) -> Double? { (v as? NSNumber)?.doubleValue }

/// Quello che il telefono sa e CarPlay mostra: la copia di `PonteAuto.kt`.
/// L'app Flutter lo aggiorna sul canale `gdanav/schermo_auto`, e dall'auto
/// tornano la ricerca, la meta scelta, le opzioni, le segnalazioni e «fine».
/// Tutto sul filo principale, come i canali di Flutter.
final class PonteAuto {
    static let shared = PonteAuto()

    var stileChiaro: String?
    var stileScuro: String?

    /// I GeoJSON delle sorgenti dello stile, come li usa la mappa del telefono.
    private(set) var sorgenti: [String: String] = [:]

    /// Dove sei: centro della mappa, e la rotta in gradi.
    private(set) var qui: (lat: Double, lon: Double)?
    private(set) var rotta = 0.0

    /// La rotta del segnaposto (la mappa guarda un po' avanti) e la sua immagine.
    private(set) var rottaIo = 0.0
    private(set) var icona = "freccia"
    private(set) var guida: GuidaAuto?
    private(set) var luoghi: [LuogoAuto] = []
    private(set) var cruscotto = CruscottoAuto()
    private(set) var avviso = AvvisoAuto()

    /// Le opzioni del percorso e la voce, per il menu.
    private(set) var opzioni: [String: Any] = [:]

    /// Le viste degli svincoli, disegnate dall'app: le ultime, per manovra.
    private(set) var svincoli: [Int: UIImage] = [:]

    /// Le icone delle segnalazioni e dei punti, disegnate dall'app.
    private(set) var immagini: [String: UIImage] = [:]

    /// Cosa dire quando non si guida: «Calcolo il percorso…», o cosa non va.
    private(set) var messaggio: String?

    /// Cresce quando cambia qualcosa dei modelli di CarPlay (manovra, messaggi,
    /// luoghi, opzioni): il resto (posizione, cruscotto) ridisegna solo la mappa.
    private(set) var versioneModello = 0

    /// gdanav Premium: l'app lo dice, l'auto lo ricorda anche a telefono spento.
    var premium: Bool { UserDefaults.standard.bool(forKey: "gdanav.premium") }

    func casa() -> LuogoAuto? { luoghi.first { $0.tipo == "casa" } }
    func lavoro() -> LuogoAuto? { luoghi.first { $0.tipo == "lavoro" } }

    private var canale: FlutterMethodChannel?
    private var ascoltatori: [UUID: () -> Void] = [:]
    private var avvisoInArrivo = false

    private static let soloMappa: Set<String> = ["posizione", "sorgenti", "cruscotto", "immagini", "stili"]

    /// Le icone disegnate dall'app sono a 3×, come sul telefono.
    private static let scalaImmagini: CGFloat = 3

    func collega(registrar: FlutterPluginRegistrar) {
        let c = FlutterMethodChannel(name: "gdanav/schermo_auto", binaryMessenger: registrar.messenger())
        c.setMethodCallHandler { [weak self] call, risultato in
            self?.gestisci(call)
            risultato(nil)
        }
        canale = c
    }

    @discardableResult
    func ascolta(_ f: @escaping () -> Void) -> UUID {
        let id = UUID()
        ascoltatori[id] = f
        return id
    }

    func smetti(_ id: UUID) {
        ascoltatori.removeValue(forKey: id)
    }

    private func gestisci(_ call: FlutterMethodCall) {
        let a = call.arguments as? [String: Any] ?? [:]
        let prima = avviso.ancoraId
        switch call.method {
        case "stili":
            stileChiaro = a["chiaro"] as? String
            stileScuro = a["scuro"] as? String
        case "sorgenti":
            if let dati = a["dati"] as? [String: String] {
                sorgenti.merge(dati) { _, nuovo in nuovo }
            }
        case "posizione":
            if let lat = numero(a["lat"]), let lon = numero(a["lon"]) {
                qui = (lat, lon)
            } else {
                qui = nil
            }
            rotta = numero(a["rotta"]) ?? 0
            rottaIo = numero(a["rotta_io"]) ?? rotta
            if let i = a["icona"] as? String { icona = i }
        case "luoghi":
            luoghi = (a["elenco"] as? [[String: Any]] ?? []).compactMap { LuogoAuto($0) }
        case "messaggio":
            messaggio = a["testo"] as? String
        case "cruscotto":
            cruscotto = CruscottoAuto(
                elettrica: (a["elettrica"] as? Bool) != false,
                batteria: numero(a["batteria"]),
                autonomiaKm: numero(a["autonomia_km"]),
                velocita: numero(a["velocita"]),
                limite: numero(a["limite"]).map { Int($0) },
                arrivoBatteria: numero(a["arrivo_batteria"]),
                sostaNome: a["sosta_nome"] as? String,
                sostaKm: numero(a["sosta_km"]),
                sostaBatteria: numero(a["sosta_batteria"]),
                meteoTemperatura: numero(a["meteo_temperatura"]),
                meteoEmoji: a["meteo_emoji"] as? String,
                meteoDove: a["meteo_dove"] as? String
            )
        case "avviso":
            let prossimo = AvvisoAuto(
                titolo: a["titolo"] as? String,
                tipo: a["tipo"] as? String,
                metri: numero(a["metri"]),
                limite: numero(a["limite"]).map { Int($0) },
                ancoraId: a["ancora_id"] as? String,
                ancoraTesto: a["ancora_testo"] as? String
            )
            // Un avviso nuovo (non la stessa segnalazione un po' più vicina):
            // CarPlay lo mostra come avviso di navigazione.
            avvisoInArrivo = prossimo.titolo != nil && (prossimo.tipo != avviso.tipo || avviso.titolo == nil)
            avviso = prossimo
        case "opzioni":
            opzioni = a
        case "svincolo":
            if let id = numero(a["id"]).map({ Int($0) }),
               let png = (a["png"] as? FlutterStandardTypedData)?.data,
               let immagine = UIImage(data: png, scale: 2)
            {
                // Solo le ultime: una vista pesa.
                let tenute = svincoli.keys.sorted().suffix(2)
                svincoli = svincoli.filter { tenute.contains($0.key) }
                svincoli[id] = immagine
            }
        case "immagini":
            for (nome, dati) in a["png"] as? [String: FlutterStandardTypedData] ?? [:] {
                if let immagine = UIImage(data: dati.data, scale: PonteAuto.scalaImmagini) {
                    immagini[nome] = immagine
                }
            }
        case "premium":
            UserDefaults.standard.set((a["sbloccato"] as? Bool) == true, forKey: "gdanav.premium")
        case "guida":
            guida = (a["attiva"] as? Bool) == true ? guidaDa(a) : nil
        default:
            return
        }
        if !PonteAuto.soloMappa.contains(call.method) && (call.method != "avviso" || avviso.ancoraId != prima) {
            versioneModello += 1
        }
        avvisa()
    }

    private func guidaDa(_ a: [String: Any]) -> GuidaAuto {
        let arrivoMs = numero(a["arrivo"]) ?? Date().timeIntervalSince1970 * 1000
        return GuidaAuto(
            tipo: numero(a["tipo"]).map { Int($0) } ?? 8,
            distanzaM: numero(a["distanza"]) ?? 0,
            strada: a["strada"] as? String ?? "",
            istruzione: a["istruzione"] as? String ?? "",
            restantiM: numero(a["restanti"]) ?? 0,
            restantiS: numero(a["secondi"]) ?? 0,
            arrivo: Date(timeIntervalSince1970: arrivoMs / 1000),
            destinazione: a["destinazione"] as? String ?? "",
            destinazioneLat: numero(a["destinazione_lat"]),
            destinazioneLon: numero(a["destinazione_lon"]),
            arrivoBatteria: numero(a["arrivo_batteria"]),
            uscita: a["uscita"] as? String ?? "",
            verso: a["verso"] as? String ?? "",
            rotonda: numero(a["rotonda"]).map { Int($0) },
            corsie: (a["corsie"] as? [[String: Any]] ?? []).map { c in
                CorsiaAuto(
                    direzioni: (c["direzioni"] as? [Any] ?? []).map { "\($0)" },
                    giusta: (c["giusta"] as? Bool) == true,
                    consigliata: c["consigliata"] as? String
                )
            },
            dopoTipo: numero(a["dopo_tipo"]).map { Int($0) },
            dopoStrada: a["dopo_strada"] as? String ?? "",
            svincolo: numero(a["svincolo"]).map { Int($0) }
        )
    }

    /// L'avviso nuovo da mostrare, una volta sola.
    func prendiAvvisoNuovo() -> AvvisoAuto? {
        guard avvisoInArrivo else { return nil }
        avvisoInArrivo = false
        return avviso
    }

    /// Qualcosa è cambiato sull'auto senza passare dal telefono (la vista 3D):
    /// gli schermi aperti si rifanno.
    func rinfresca() {
        versioneModello += 1
        avvisa()
    }

    private func avvisa() {
        for f in ascoltatori.values { f() }
    }

    // MARK: - Le domande all'app

    /// Una domanda all'app, con la risposta sul filo principale.
    func chiedi(_ metodo: String, _ argomenti: Any? = nil, risposta: @escaping (Any?) -> Void = { _ in }) {
        guard let c = canale else { return risposta(nil) }
        c.invokeMethod(metodo, arguments: argomenti) { r in
            if r is FlutterError || (r as? NSObject) === FlutterMethodNotImplemented {
                risposta(nil)
            } else {
                risposta(r)
            }
        }
    }

    /// La ricerca fatta sull'auto, con Photon come sul telefono.
    func cerca(_ testo: String, risultati: @escaping ([LuogoAuto]) -> Void) {
        chiedi("cerca", ["testo": testo]) { r in
            risultati((r as? [[String: Any]] ?? []).compactMap { LuogoAuto($0) })
        }
    }

    /// Una meta scelta sull'auto: il telefono calcola e parte la guida.
    func vai(_ l: LuogoAuto) {
        messaggio = "Calcolo il percorso per \(l.etichetta)…"
        versioneModello += 1
        avvisa()
        chiedi("vai", l.comeMappa)
    }

    /// In guida si passa dal distributore e poi si prosegue; fermi, ci si va.
    func passa(_ l: LuogoAuto) {
        messaggio = "Passo da \(l.etichetta)…"
        versioneModello += 1
        avvisa()
        chiedi("passa", l.comeMappa)
    }

    /// Casa o Lavoro scelti dalla ricerca sull'auto.
    func imposta(_ tipo: String, _ l: LuogoAuto, fatto: @escaping (Bool) -> Void) {
        chiedi("imposta", ["tipo": tipo, "luogo": l.comeMappa]) { fatto(($0 as? Bool) == true) }
    }

    func cambiaOpzione(_ chiave: String, _ valore: Any) {
        chiedi("opzioni", [chiave: valore])
    }

    func alternaVoce() {
        chiedi("voce")
    }

    /// Segnala dove sei; la risposta è la frase da mostrare.
    func segnala(_ tipo: String, risposta: @escaping (String) -> Void) {
        chiedi("segnala", ["tipo": tipo]) { risposta($0 as? String ?? "Non è partita.") }
    }

    func ancora(_ id: String, si: Bool) {
        chiedi("ancora", ["id": id, "si": si])
    }

    /// Le colonnine rapide vicine, come luoghi da raggiungere.
    func colonnine(risultati: @escaping ([LuogoAuto]) -> Void) {
        chiedi("colonnine") { r in
            risultati((r as? [[String: Any]] ?? []).compactMap { LuogoAuto($0, tipo: "colonnina") })
        }
    }

    /// Auto termica: i distributori intorno.
    func distributori(risultati: @escaping ([LuogoAuto]) -> Void) {
        chiedi("distributori") { r in
            risultati((r as? [[String: Any]] ?? []).compactMap { LuogoAuto($0, tipo: "distributore") })
        }
    }

    /// L'auto ha chiesto di finire la guida.
    func fermaDallAuto() {
        chiedi("ferma")
    }
}
